pkg/codegen/go/gen_program_expressions.go

```diff
@@ -375,20 +496,84 @@ func (g *generator) argumentTypeName(expr model.Expression, destType model.Type,
 		}
 	}
 
-	// TODO support rest of types
 	switch destType := destType.(type) {
 	case *model.OpaqueType:
 		switch destType {
+		case model.IntType:
+			if isInput {
+				return "pulumi.Int"
+			}
+			return "int"
 		case model.NumberType:
+			if isInput {
+				return "pulumi.Float64"
+			}
 			return "float64"
+		case model.StringType:
+			if isInput {
+				return "pulumi.String"
+			}
+			return "string"
+		case model.BoolType:
+			if isInput {
+				return "pulumi.Bool"
+			}
+			return "bool"
 		default:
 			return destType.Name
 		}
-	// TODO could probably improve these types by inspecting kv types
-	case *model.ObjectType, *model.MapType:
+	case *model.ObjectType:
+		return "map[string]interface{}"
+	case *model.MapType:
+		valType := argumentTypeName(nil, destType.ElementType, isInput)
+		return fmt.Sprintf("map[string]%s", valType)
+	case *model.ListType:
+		argTypeName := argumentTypeName(nil, destType.ElementType, isInput)
+		if strings.HasPrefix(argTypeName, "pulumi.") {
+			return fmt.Sprintf("%sArray", argTypeName)
+		}
+		return fmt.Sprintf("[]%s", argTypeName)
+	case *model.TupleType:
+		// attempt to collapse tuple types
+		var elmType model.Type
+		for i, t := range destType.ElementTypes {
+			if i == 0 {
+				elmType = t
+			}
+
+			if !elmType.Equals(t) {
+				elmType = nil
+				break
+			}
+		}
+
+		if elmType != nil {
+			argTypeName := argumentTypeName(nil, elmType, isInput)
+			if strings.HasPrefix(argTypeName, "pulumi.") {
+				return fmt.Sprintf("%sArray", argTypeName)
+			}
+			return fmt.Sprintf("[]%s", argTypeName)
+		}
+
+		if isInput {
+			return "pulumi.Array"
+		}
+		return "[]interface{}"
+	case *model.OutputType:
+		isInput = true
+		return argumentTypeName(expr, destType.ElementType, isInput)
+	case *model.UnionType:
+		for _, ut := range destType.ElementTypes {
+			switch ut.(type) {
+			case *model.OpaqueType:
+				return argumentTypeName(expr, ut, isInput)
```

Comment:
```suggestion
			if _, isOpaqueType := ut.(*model.OpaqueType); isOpaqueType {
				return argumentTypeName(expr, ut, isInput)
```

Reply:
Also, what case is this intended to handle?

Reply:
`__convert`. The diff if removed:

```
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -31,4 +31,4 @@
            	            	 					Protocol: pulumi.String("-1"),
            	            	-					FromPort: pulumi.Int(0),
            	            	-					ToPort:   pulumi.Int(0),
            	            	+					FromPort: interface{}(0),
            	            	+					ToPort:   interface{}(0),
            	            	 					CidrBlocks: pulumi.StringArray{
            	            	@@ -41,4 +41,4 @@
            	            	 					Protocol: pulumi.String("tcp"),
            	            	-					FromPort: pulumi.Int(80),
            	            	-					ToPort:   pulumi.Int(80),
            	            	+					FromPort: interface{}(80),
            	            	+					ToPort:   interface{}(80),
            	            	 					CidrBlocks: pulumi.StringArray{
            	            	@@ -96,3 +96,3 @@
            	            	 		webTargetGroup, err := elasticloadbalancingv2.NewTargetGroup(ctx, "webTargetGroup", &elasticloadbalancingv2.TargetGroupArgs{
            	            	-			Port:       pulumi.Int(80),
            	            	+			Port:       interface{}(80),
            	            	 			Protocol:   pulumi.String("HTTP"),
            	            	@@ -106,3 +106,3 @@
            	            	 			LoadBalancerArn: webLoadBalancer.Arn,
            	            	-			Port:            pulumi.Int(80),
            	            	+			Port:            interface{}(80),
            	            	 			DefaultActions: elasticloadbalancingv2.ListenerDefaultActionArray{
            	            	@@ -150,3 +150,3 @@
            	            	 			Cluster:        cluster.Arn,
            	            	-			DesiredCount:   pulumi.Int(5),
            	            	+			DesiredCount:   interface{}(5),
            	            	 			LaunchType:     pulumi.String("FARGATE"),
            	            	@@ -164,3 +164,3 @@
            	            	 					ContainerName:  pulumi.String("my-app"),
            	            	-					ContainerPort:  pulumi.Int(80),
            	            	+					ContainerPort:  interface{}(80),
            	            	 				},
```

Reply:
Ah, okay. This happens because `RewriteConversions` doesn't pick the best dest type out of the available types in a union. Should this just recurse in all cases (not just on `OpaqueType`)?

Reply:
Probably, but I'd prefer to wait until I have an example or test case to develop against though. I'm sure I'll end up hitting this when I go to match one of the three remaining test cases or via a bug report.